Controlled upgrades

dpsutton opened the pull request:
Let's add a `rollout` number 0-100 on `latest`


Version information has two keys:

```clojure
{:latest {:version "v1.50.25",
          :released "2024-09-10",
          :patch true,
          :highlights [,,,]},
 :older [{:version "v1.50.24",
          :released "2024-09-03",
          :patch true,
          :highlights [,,,]}]}
```

The proposal here is that we add a `rollout` to the `latest` map, and then we only show the `latest` to the frontend (and therefore the upgrade) if some number of the local instance is lower than the rollout number.

```clojure
{:latest {:version "v1.50.25",
          :released "2024-09-10",
          :rollout 50 ;; number between 0 - 100 <--------- new
          :patch true,
          :highlights [,,,]},
 :older [{:version "v1.50.24",
          :released "2024-09-03",
          :patch true,
          :highlights [,,,]}]}
```

<img width="1247" alt="image" src="https://github.com/user-attachments/assets/7c0e18c3-b609-4b80-a95b-14445b58b167">


This derives the rollout threshold of the instance from the `site-uuid` which is constant for all instances in a cluster.

```clojure
(defsetting upgrade-threshold
  (deferred-tru "Threshold (value in 0-100) indicating at which treshold it should offer an upgrade to the latest major version.")
  :visibility :internal
  :export?    false
  :type       :integer
  :setter     :none
  :getter     (fn []
                (-> (site-uuid) hash (mod 100))))
```

(and of course overridable with `MB_UPGRADE_THRESHOLD=90 java -jar ...` if they want to be in a higher cohort)

This lets us uniformly distribute self-hosted instances into buckets of 0-100. And as we increase the `rollout` number of the version-info json file, more instances will show the upgrade information in the application.


```shell
❯ http get "http://localhost:3000/api/setting/version-info" x-api-key:$API_KEY | jq 'keys'
[
  "older"
]

## vs when it should show
❯ http get "http://localhost:3000/api/setting/version-info" x-api-key:$API_KEY | jq 'keys'
[
  "latest",
  "older"
]
```

This really only matters across major upgrades. We always show upgrades of minor upgrades in the same major.
Ex: `0.49.5` will always show a banner for a latest of `0.49.6`
Of course `0.49.5` will compare rollout thresholds on whether to show an update to the latest of `0.50.2`
Also, `0.49.5` will compare rollout thresholds on whether to show an update to the latest of `0.51.2` (*2* majors). We can make this show the latest 50 in the future if we want. (and the same scenario for more exaggerated version jumps, 49 -> 83 ,etc).

See [this explanation](https://www.notion.so/metabase/Checking-A-feature-in-multiple-instances-0064e6a643624b1d91cb94a5221df6da) in notion of how it worked running it 200 times.

Started a jar up 200 times, fetching versions that included a 25% rollout threshold and had the following results:
<img width="738" alt="image" src="https://github.com/user-attachments/assets/c616ed13-ec6b-4974-9aba-99c361b50994">

```
site-uuid, threshold, upgrade-available, latest-version
35095708-ad3f-42b2-8e26-65b475ec5cd7, 2, true, 1.51.1
a247324a-82c1-41e1-9f91-7016a66e9014, 87, false, 
8072a554-3ded-4a4f-83ec-0d363e7f0deb, 23, true, 1.51.1
b40db850-028e-4d17-8806-0d27758395f6, 67, false, 
a466def7-fa50-4cf7-a2ee-8b31b422749a, 20, true, 1.51.1
5effa39b-9065-49ca-b686-f17eaecf4198, 85, false, 
4158efe4-3703-4372-b94a-9458f975fa7c, 55, false, 
```

dpsutton committed:
Initial commit of controlled upgrades

- new setting upgrade-threshold (`MB_UPGRADE_THRESHOLD`)
  number 0-100
- conditionally removing latest from upgrade checks

AS OF RIGHT NOW IT ALWAYS REMOVES the latest.
Should help FE make this optional

dpsutton committed:
dumb mistake

iethree committed:
handle no latest version info

dpsutton committed:
implement and basic test

dpsutton committed:
More tests, refactor name

dpsutton committed:
more tests

dpsutton committed:
cljfmt does not like commas in source code.

noahmoss commented on src/metabase/public_settings.clj:
```diff
@@ -70,12 +70,54 @@
   :audit   :getter
   :default true)
 
+(declare site-uuid)
```
Maybe just move the `site-uuid` setting (+ related settings) up to the top? I don't think there's really a specific order right now

dpsutton replied:
done. good point. I think i was trying to minimize the diff at first, which is not necessary at all.

dpsutton committed:
Just move site-uuid to the declaration and don't declare it

nothing is sacred in the settings order. no reason. i think i was just
trying to minimize the diff, which a dumb and ignoble goal.

rafpaf commented on frontend/src/metabase/admin/settings/components/SettingsUpdatesForm/VersionUpdateNotice/VersionUpdateNotice.tsx:
```diff
@@ -52,11 +47,17 @@ function OnLatestVersion({ currentVersion }) {
+  return (
+    <div>
+      <OnLatestVersionMessage>
+        {t`You're running Metabase ${currentVersion}`}
```
Context string suggestion: `c("{0} is a version number."}`

iethree committed:
add context strings

iethree committed:
update unit tests

iethree committed:
use display version

piranha commented on src/metabase/public_settings.clj:
```diff
@@ -70,12 +70,60 @@
+  :type       :integer
+  :setter     :none
+  :getter     (fn []
+                (-> (site-uuid) hash (mod 100))))
```
I mentioned that in Slack, but maybe we don't want this number to be THAT stable and adding a version info (or something like this) to re-hash our distribution after every upgrade is the better way?

dpsutton replied:
updated! good idea

dpsutton committed:
Have upgrade threshold include current major version

hopefully helps rotate people from early to late, and late to early
across major versions. Idea from sanya and quite nice!
